datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java

```diff
@@ -1,173 +1,297 @@
+        return new AbstractSdxAction<>(SdxDetachFailedEvent.class) {
             @Override
-            protected SdxContext createFlowContext(FlowParameters flowParameters, StateContext<FlowState, FlowEvent> stateContext,
-                    SdxDetachSuccessEvent payload) {
+            protected SdxContext createFlowContext(FlowParameters flowParameters,
+                    StateContext<FlowState, FlowEvent> stateContext,
+                    SdxDetachFailedEvent payload) {
                 return SdxContext.from(flowParameters, payload);
             }
 
             @Override
-            protected void doExecute(SdxContext context, SdxDetachSuccessEvent payload, Map<Object, Object> variables) throws Exception {
-                LOGGER.info("SDX detach finalized: {}", payload.getResourceId());
-                sdxStatusService.setStatusForDatalakeAndNotify(DatalakeStatusEnum.STOPPED, "Datalake is detached", payload.getResourceId());
-                if (variables.containsKey(RESIZED_SDX)) {
-                    SdxCluster sdxCluster = (SdxCluster) variables.get(RESIZED_SDX);
-                    MDCBuilder.buildMdcContext(sdxCluster);
-                    try {
-                        transactionService.required(() -> {
-                            SdxCluster created = sdxClusterRepository.save(sdxCluster);
-                            ownerAssignmentService.assignResourceOwnerRoleIfEntitled(created.getInitiatorUserCrn(), created.getCrn(), created.getAccountId());
-                            sdxStatusService.setStatusForDatalake(DatalakeStatusEnum.REQUESTED, "Data lake requested", created);
-                            context.setSdxId(created.getId());
-                            return created;
-                        });
-                        LOGGER.info("Detaching Sdx with Name: {} and Crn: {} is complete", sdxCluster.getName(), sdxCluster.getCrn());
-                    } catch (TransactionService.TransactionExecutionException e) {
-                        throw new TransactionService.TransactionRuntimeExecutionException(e);
-                    }
+            protected void doExecute(SdxContext context, SdxDetachFailedEvent payload,
+                    Map<Object, Object> variables) throws Exception {
+                SdxCluster resized = (SdxCluster) variables.get(RESIZED_SDX);
+                SdxCluster detached = (SdxCluster) variables.get(DETACHED_SDX);
+                LOGGER.error("Failed to attach new cluster with ID {} during detaching of cluster with ID {}." +
+                        " Attempting to reattach detached cluster.", resized.getId(), detached.getId());
+
+                // Reattach detached cluster.
+                detached = sdxAttachService.reattachSdx(detached);
+
+                // Remove new cluster from repository in case it was added.
+                if (sdxClusterRepository.existsById(resized.getId())) {
+                    sdxClusterRepository.delete(resized);
```

Comment:
Why do you think this logic is needed?

with the below code even in case of failure from assignResourceOwnerRoleIfEntitled, the repository update will be reverted.
```
 return transactionService.required(() -> {
            SdxCluster created = sdxClusterRepository.save(cluster);
            ownerAssignmentService.assignResourceOwnerRoleIfEntitled(
                    created.getInitiatorUserCrn(), created.getCrn(), created.getAccountId()
            );
            return created;
        });
```

Reply:
I am relatively unfamiliar with the transactionService. So if any step in the code fails there the block will essentially not be run?
I added those lines in from the conversation we had with Dan where he mentioned that we may want to remove the new cluster from the repository.

Reply:
Let's say there is an exception from assignResourceOwnerRoleIfEntitled. Update to the repository would automatically be rolled back.

You can test it by injecting a failure in the transaction block, I updated in my comment above.

Reply:
Perfect this makes sense to me thank you, I will remove this.

Reply:
Done.